src/Infrastructure.VS/IVsVersionProvider.cs

```diff
@@ -0,0 +1,83 @@
+﻿/*
+ * SonarLint for Visual Studio
+ * Copyright (C) 2016-2021 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System;
+using System.ComponentModel.Composition;
+using Microsoft.VisualStudio.Shell;
+using Microsoft.VisualStudio.Shell.Interop;
+
+namespace SonarLint.VisualStudio.Infrastructure.VS
+{
+    public interface IVsVersionProvider
+    {
+        IVsVersion Version { get; }
+    }
+
+    [Export(typeof(IVsVersionProvider))]
+    [PartCreationPolicy(CreationPolicy.Shared)]
+    internal class VsVersionProvider : IVsVersionProvider
+    {
+        public IVsVersion Version { get; }
+
+        [ImportingConstructor]
+        public VsVersionProvider([Import(typeof(SVsServiceProvider))] IServiceProvider serviceProvider)
+        {
+            var vsShell = serviceProvider.GetService(typeof(SVsShell)) as IVsShell;
+
+            vsShell.GetProperty((int)__VSSPROPID5.VSSPROPID_AppBrandName, out var name);
+            vsShell.GetProperty((int)__VSSPROPID5.VSSPROPID_AppShortBrandName, out var shortName);
+            vsShell.GetProperty((int)__VSSPROPID5.VSSPROPID_ReleaseVersion, out var version);
+
+            Version = new VsVersion((string)name, (string)shortName, (string)version);
+        }
+    }
+
+    public interface IVsVersion
+    {
+        /// <summary>
+        /// See <see cref="__VSSPROPID5.VSSPROPID_AppBrandName"/>
+        /// </summary>
+        string Name { get; }
+
+        /// <summary>
+        /// See <see cref="__VSSPROPID5.VSSPROPID_AppShortBrandName"/>
+        /// </summary>
+        string ShortName { get; }
+
+        /// <summary>
+        /// See <see cref="__VSSPROPID5.VSSPROPID_ReleaseVersion"/>
+        /// </summary>
+        string BuildVersion { get; }
```

Comment:
I investigated further the values that are returned from this property.

* In 2015 it's `14.0.25420.1 D14REL` when the about window says `14.0.25431.01 update 3`
* In 2017 it's `15.0.28307.1259 D15.9` when the about window says `15.9.27`
* In 2019 it's `16.0.30914.41 D16.9` when the about window says `Version 16.9.0 Preview 3.0`.

I was able to cross reference the number with the release versions of [2019](https://docs.microsoft.com/en-us/visualstudio/install/visual-studio-build-numbers-and-release-dates?view=vs-2019) and [2017](https://docs.microsoft.com/en-us/visualstudio/install/visual-studio-build-numbers-and-release-dates?view=vs-2017&preserve-view=true) (couldn't find for 2015). 
Notice that the minor field does not match: in both 2019 and 2017 it seems to always be zero:
* `16.0.30914.41 D16.9` matches `16.9.30914.41`
* `15.0.28307.1259 D15.9` matches `15.9.28307.1259`

I tried to cross-reference it with the values that we already have in telemetry -- those don't match. Some of them do, but for instance the telemetry reports `16.0.30907.39` which is nowhere to be found in the release table. So I think that what we're currently sending to the telemetry is definitely not enough and we should be sending this field instead.

The question is: what should we send to telemetry? 
I suppose it would be ok to send the field as-is; in most cases we would be interested in knowing "who is below update x" -- which would require us to go to the release notes, get the number, and then filter telemetry for anyone below that number.
Alternatively, we could "patch" this field and take the minor version from whatever is reported in the last part (e.g. the `D15.9` from `15.0.28307.1259 D15.9`). But that feels kinda hacky.

I think we should just send it as-is. WDYT?

Reply:
That's very irritating. At least there are official pages that list the versions for VS2017 and 2019, but it's still going to be a bit of a pain to have to cross-reference them.

I'll have a quick look to see if I can find an alternative, but I'm not expecting to find anything you haven't.